### PR TITLE
[meson] fix cross compiling from linux to windows

### PR DESCRIPTION
**Pull request by autoantwort**
Fixes cross comping from linux to windows, you currently get [this](https://github.com/mesonbuild/meson/issues/8647) error

### CODE REVIEW

File: scripts/cmake/vcpkg_configure_meson.cmake [hunk 1]
```diff
@@ -58,7 +58,7 @@ function(vcpkg_internal_meson_generate_native_file _additional_binaries) #https:
             string(APPEND NATIVE "${proglower} = '${VCPKG_DETECTED_CMAKE_${prog}_COMPILER}'\n")
         endif()
     endforeach()
-    if(VCPKG_DETECTED_CMAKE_LINKER AND VCPKG_TARGET_IS_WINDOWS)
+    if(VCPKG_DETECTED_CMAKE_LINKER AND VCPKG_TARGET_IS_WINDOWS AND NOT VCPKG_TARGET_IS_MINGW)
```
Comment (Neumann-A) [inline scripts/cmake/vcpkg_configure_meson.cmake:61]: I think the real check is if the compiler is gcc|g++|icc
`AND NOT <COMPILER_VAR> MATCHES (gcc|g++|icc)`
maybe we can get the compiler_id from cmake for this check?

the reason for this is that meson is implemented in such a way that that it will pass c(pp)_ld as -fuse_ld=<c(pp)_ld> to the compiler which works with clang since it is re-targetable but breaks with gcc|icc.

**Comment by autoantwort** [comment]
`gdk-pixbuf` fails because of 
```
CMake Error at scripts/cmake/vcpkg_copy_tools.cmake:59 (message):
  Couldn't find this tool:
  D:/packages/gdk-pixbuf_x86-windows/bin/gdk-pixbuf-thumbnailer.exe.
Call Stack (most recent call first):
  ports/gdk-pixbuf/portfile.cmake:67 (vcpkg_copy_tools)
  scripts/ports.cmake:140 (include)

```

and `fribidi` because of
```
..\src\3bb089f553-744296bd76.clean\gen.tab\meson.build:32:0: ERROR: No build machine compiler for "gen.tab\gen-unicode-version.c"
```
seems that it should have a host dependency or we set the host compiler or we don't use `VCPKG_CROSSCOMPILING` on windows

File: scripts/cmake/vcpkg_configure_meson.cmake (hunk 1 above)
Comment (autoantwort) [reply scripts/cmake/vcpkg_configure_meson.cmake]: Seems that it must be ` AND NOT <COMPILER_VAR> MATCHES "(gcc|g\\+\\+|icc)"`

File: scripts/cmake/vcpkg_configure_meson.cmake [hunk 2]
```diff
@@ -316,7 +320,7 @@ function(vcpkg_internal_meson_generate_cross_file _additional_binaries) #https:/
         string(APPEND CROSS "cpu = '${BUILD_CPU}'\n")
     endif()
 
-    if(NOT BUILD_CPU_FAM MATCHES "${HOST_CPU_FAM}" OR VCPKG_TARGET_IS_ANDROID OR VCPKG_TARGET_IS_IOS OR VCPKG_TARGET_IS_UWP)
+    if(NOT BUILD_CPU_FAM MATCHES "${HOST_CPU_FAM}" OR VCPKG_CROSSCOMPILING)
```
Comment (dg0yt) [inline scripts/cmake/vcpkg_configure_meson.cmake:323]: With ` OR VCPKG_TARGET_IS_MINGW` added, it also fixes building ports like harfbuzz with mingw on Windows.

Comment (autoantwort) [reply scripts/cmake/vcpkg_configure_meson.cmake:323]: Is for `VCPKG_TARGET_IS_MINGW` `VCPKG_CROSSCOMPILING` not also true, or do you explicitly set the host triplet to mingw?

Comment (dg0yt) [reply scripts/cmake/vcpkg_configure_meson.cmake:323]: When I do not have MSVC, I must set the host triplet to mingw. I can't even install `vcpkg-cmake` without this.

Comment (autoantwort) [reply scripts/cmake/vcpkg_configure_meson.cmake:323]: Ok, then it sounds like adding `OR VCPKG_TARGET_IS_MINGW` makes more sense than using `VCPKG_CROSSCOMPILING`

File: scripts/get_cmake_vars/CMakeLists.txt [hunk 3]
```diff
@@ -18,7 +18,7 @@ set(PROGLIST AR RANLIB STRIP NM OBJDUMP DLLTOOL MT LINKER)
 foreach(prog IN LISTS PROGLIST)
     list(APPEND VCPKG_DEFAULT_VARS_TO_CHECK CMAKE_${prog})
 endforeach()
-set(COMPILERS ${VCPKG_LANGUAGES} RC)
+set(COMPILERS ${VCPKG_LANGUAGES} RC C CXX)
```
Comment (Neumann-A) [inline scripts/get_cmake_vars/CMakeLists.txt:21]: revert this change it is unnecessary. VCPKG_LANGUAGES is set to C and CXX

Comment (Neumann-A) [reply scripts/get_cmake_vars/CMakeLists.txt:21]: https://github.com/microsoft/vcpkg/blob/dba1f6beb92ebd05bbda4f47a23733564e04156f/scripts/get_cmake_vars/CMakeLists.txt#L3

File: scripts/cmake/vcpkg_configure_meson.cmake (hunk 2 above)
Comment (dg0yt) [reply scripts/cmake/vcpkg_configure_meson.cmake]: I spoke too soon. It compiled harfbuzz, but created .so files for x64-mingw-dynamic on Windows. Does it work better for cross-compiling from Linux?

I collected some fixes, but meson in vcpkg seems to need more work for mingw.

Comment (autoantwort) [reply scripts/cmake/vcpkg_configure_meson.cmake]: I used the static triplet and it build qtbase without a linker error, so I assume that worked.

**Comment by JackBoosY** [comment]
Seems the cla check has some problems, @autoantwort can you re-sign the cla?

**Comment by autoantwort** [comment]
> Seems the cla check has some problems, @autoantwort can you re-sign the cla?

How?

**Comment by JackBoosY** [comment]
@autoantwort  Can you commit some small changes to refresh the ci status?

**Comment by autoantwort** [comment]
> @autoantwort Can you commit some small changes to refresh the ci status?

Seems that the `git commit --allow-empty --amend` helped.

**Comment by talregev** [comment]
> `gdk-pixbuf` fails because of
> 
> ```
> CMake Error at scripts/cmake/vcpkg_copy_tools.cmake:59 (message):
>   Couldn't find this tool:
>   D:/packages/gdk-pixbuf_x86-windows/bin/gdk-pixbuf-thumbnailer.exe.
> Call Stack (most recent call first):
>   ports/gdk-pixbuf/portfile.cmake:67 (vcpkg_copy_tools)
>   scripts/ports.cmake:140 (include)
> ```
> 

I have the same error when I compile `gdk-pixbuf` on x64-mingw-dyanmic.
I will open a new issue for that.

**Review by JackBoosY — APPROVED** [review]

**Comment by talregev** [comment]
@dg0yt @Neumann-A Can you help to review this fix? 
Thank you.

File: scripts/cmake/vcpkg_configure_meson.cmake [hunk 4]
```diff
@@ -54,12 +54,15 @@ function(vcpkg_internal_meson_generate_native_file _additional_binaries) #https:
     foreach(prog IN LISTS compiler)
         if(VCPKG_DETECTED_CMAKE_${prog}_COMPILER)
             string(REPLACE "CXX" "CPP" mesonprog "${prog}")
+            string(REPLACE "RC" "windres" mesonprog "${mesonprog}") # https://mesonbuild.com/Windows-module.html
             string(TOLOWER "${mesonprog}" proglower)
             string(APPEND NATIVE "${proglower} = '${VCPKG_DETECTED_CMAKE_${prog}_COMPILER}'\n")
         endif()
     endforeach()
-    if(VCPKG_DETECTED_CMAKE_LINKER AND VCPKG_TARGET_IS_WINDOWS)
+    if(VCPKG_DETECTED_CMAKE_LINKER AND VCPKG_TARGET_IS_WINDOWS AND NOT VCPKG_DETECTED_CMAKE_C_COMPILER MATCHES "(gcc|icc)")
```
Comment (strega-nil-ms) [inline scripts/cmake/vcpkg_configure_meson.cmake:62]:
```suggestion
    if(VCPKG_DETECTED_CMAKE_LINKER AND VCPKG_TARGET_IS_WINDOWS AND NOT VCPKG_DETECTED_CMAKE_C_COMPILER MATCHES "gcc|icc")
```

File: scripts/cmake/vcpkg_configure_meson.cmake [hunk 5]
```diff
@@ -54,12 +54,15 @@ function(vcpkg_internal_meson_generate_native_file _additional_binaries) #https:
...
         string(APPEND NATIVE "c_ld = '${VCPKG_DETECTED_CMAKE_LINKER}'\n")
+    endif()
+    if(VCPKG_DETECTED_CMAKE_LINKER AND VCPKG_TARGET_IS_WINDOWS AND NOT VCPKG_DETECTED_CMAKE_CXX_COMPILER MATCHES "(g\\+\\+|icc)")
```
Comment (strega-nil-ms) [inline scripts/cmake/vcpkg_configure_meson.cmake:65]:
```suggestion
    if(VCPKG_DETECTED_CMAKE_LINKER AND VCPKG_TARGET_IS_WINDOWS AND NOT VCPKG_DETECTED_CMAKE_CXX_COMPILER MATCHES [[g\+\+|icc]]")
```

**Review by strega-nil-ms — APPROVED** [review]

File: scripts/cmake/vcpkg_configure_meson.cmake [hunk 6]
```diff
@@ -54,12 +54,15 @@ function(vcpkg_internal_meson_generate_native_file _additional_binaries) #https:
     foreach(prog IN LISTS compiler)
         if(VCPKG_DETECTED_CMAKE_${prog}_COMPILER)
             string(REPLACE "CXX" "CPP" mesonprog "${prog}")
+            string(REPLACE "RC" "windres" mesonprog "${mesonprog}") # https://mesonbuild.com/Windows-module.html
             string(TOLOWER "${mesonprog}" proglower)
             string(APPEND NATIVE "${proglower} = '${VCPKG_DETECTED_CMAKE_${prog}_COMPILER}'\n")
         endif()
     endforeach()
-    if(VCPKG_DETECTED_CMAKE_LINKER AND VCPKG_TARGET_IS_WINDOWS)
+    if(VCPKG_DETECTED_CMAKE_LINKER AND VCPKG_TARGET_IS_WINDOWS AND NOT VCPKG_DETECTED_CMAKE_C_COMPILER MATCHES "gcc|icc")
```
Comment (strega-nil-ms) [inline scripts/cmake/vcpkg_configure_meson.cmake:62]: I don't understand this; could you explain? (preferably add a comment too?)

**Review by strega-nil-ms — CHANGES_REQUESTED** [review]

File: scripts/cmake/vcpkg_configure_meson.cmake (hunk 6 above)
Comment (autoantwort) [reply scripts/cmake/vcpkg_configure_meson.cmake:62]: For gcc and ICC you can not explicitly set the linker

Comment (talregev) [reply scripts/cmake/vcpkg_configure_meson.cmake:62]: @autoantwort Can you add a comment in the code that explain this?

Comment (autoantwort) [reply scripts/cmake/vcpkg_configure_meson.cmake:62]: https://github.com/mesonbuild/meson/issues/8647#issuecomment-878673456  
I can do it next week

Comment (talregev) [reply scripts/cmake/vcpkg_configure_meson.cmake:62]: ok :)
Thank you!

Comment (BillyONeal) [reply scripts/cmake/vcpkg_configure_meson.cmake:62]: It seems like the correct fix is to get the program name and pass that here rather than ignoring the setting completely?

**Comment by BillyONeal** [comment]
Dropping "reviewed" because there is outstanding discussion.

File: scripts/cmake/vcpkg_configure_meson.cmake (hunk 6 above)
Comment (autoantwort) [reply scripts/cmake/vcpkg_configure_meson.cmake:62]: When I understand it correctly you select the linker backend by adding a compiler flag and not by changing the linker executable. So the linker executeable is always the same and we can ignore the Program name. And the flag should be passed by VCPKG_LINKER_FLAGS or so

Comment (BillyONeal) [reply scripts/cmake/vcpkg_configure_meson.cmake:62]: I guess, to put it another way, the right thing to do when we don't know how to attach a user setting isn't "just ignore the setting". Either we should make it work, or check that it's what we expect and fail if it isn't.

Comment (autoantwort) [reply scripts/cmake/vcpkg_configure_meson.cmake:62]: > check that it's what we expect and fail if it isn't.

Ok I will do that.

File: scripts/cmake/vcpkg_configure_meson.cmake [hunk 7]
```diff
@@ -54,13 +54,24 @@ function(vcpkg_internal_meson_generate_native_file _additional_binaries) #https:
...
             string(APPEND NATIVE "${proglower} = '${VCPKG_DETECTED_CMAKE_${prog}_COMPILER}'\n")
         endif()
     endforeach()
     if(VCPKG_DETECTED_CMAKE_LINKER AND VCPKG_TARGET_IS_WINDOWS)
-        string(APPEND NATIVE "c_ld = '${VCPKG_DETECTED_CMAKE_LINKER}'\n")
-        string(APPEND NATIVE "cpp_ld = '${VCPKG_DETECTED_CMAKE_LINKER}'\n")
+        if (NOT VCPKG_DETECTED_CMAKE_C_COMPILER MATCHES "gcc|icc")
+            string(APPEND NATIVE "c_ld = '${VCPKG_DETECTED_CMAKE_LINKER}'\n")
+        elseif (NOT VCPKG_DETECTED_CMAKE_C_COMPILER STREQUAL VCPKG_DETECTED_CMAKE_LINKER)
+            message(FATAL_ERROR "The linker and compiler executeable must be the same for gcc and icc. Use -fuse-ld=... to change the linker backend.")
```
Comment (strega-nil-ms) [inline scripts/cmake/vcpkg_configure_meson.cmake:66]:
```suggestion
            message(FATAL_ERROR "The linker and compiler executable must be the same for gcc and icc. Use -fuse-ld=... to change the linker backend.")
```

File: scripts/cmake/vcpkg_configure_meson.cmake [hunk 8]
```diff
@@ -54,13 +54,24 @@ function(vcpkg_internal_meson_generate_native_file _additional_binaries) #https:
...
+        endif()
+    endif()
+    if(VCPKG_DETECTED_CMAKE_LINKER AND VCPKG_TARGET_IS_WINDOWS)
+        if (NOT VCPKG_DETECTED_CMAKE_CXX_COMPILER MATCHES [[g\+\+|icc]])
+            string(APPEND NATIVE "cpp_ld = '${VCPKG_DETECTED_CMAKE_LINKER}'\n")
+        elseif (NOT VCPKG_DETECTED_CMAKE_CXX_COMPILER STREQUAL VCPKG_DETECTED_CMAKE_LINKER)
+            message(FATAL_ERROR "The linker and compiler executeable must be the same for gcc and icc. Use -fuse-ld=... to change the linker backend.")
```
Comment (strega-nil-ms) [inline scripts/cmake/vcpkg_configure_meson.cmake:73]:
```suggestion
            message(FATAL_ERROR "The linker and compiler executable must be the same for gcc and icc. Use -fuse-ld=... to change the linker backend.")
```

File: scripts/cmake/vcpkg_configure_meson.cmake [hunk 9]
```diff
@@ -79,12 +90,14 @@ function(vcpkg_internal_meson_generate_native_file _additional_binaries) #https:
...
         list(TRANSFORM WIN_C_STANDARD_LIBRARIES APPEND "'")
         list(TRANSFORM WIN_C_STANDARD_LIBRARIES PREPEND "'")
+        list(FILTER WIN_C_STANDARD_LIBRARIES EXCLUDE REGEX "^''$")
         list(JOIN WIN_C_STANDARD_LIBRARIES ", " WIN_C_STANDARD_LIBRARIES)
         string(APPEND NATIVE "c_winlibs = [${WIN_C_STANDARD_LIBRARIES}]\n")
         string(REGEX REPLACE "( |^)(-|/)" ";\\2" WIN_CXX_STANDARD_LIBRARIES "${VCPKG_DETECTED_CMAKE_CXX_STANDARD_LIBRARIES}")
         string(REGEX REPLACE "\\.lib " ".lib;" WIN_CXX_STANDARD_LIBRARIES "${WIN_CXX_STANDARD_LIBRARIES}")
         list(TRANSFORM WIN_CXX_STANDARD_LIBRARIES APPEND "'")
         list(TRANSFORM WIN_CXX_STANDARD_LIBRARIES PREPEND "'")
+        list(FILTER WIN_CXX_STANDARD_LIBRARIES EXCLUDE REGEX "^''$")
```
Comment (strega-nil-ms) [inline scripts/cmake/vcpkg_configure_meson.cmake:100]:
```suggestion
        list(REMOVE_ITEM WIN_CXX_STANDARD_LIBRARIES "''")
```

File: scripts/cmake/vcpkg_configure_meson.cmake [hunk 10]
```diff
@@ -79,12 +90,14 @@ function(vcpkg_internal_meson_generate_native_file _additional_binaries) #https:
         string(REGEX REPLACE "\\.lib " ".lib;" WIN_C_STANDARD_LIBRARIES "${WIN_C_STANDARD_LIBRARIES}")
         list(TRANSFORM WIN_C_STANDARD_LIBRARIES APPEND "'")
         list(TRANSFORM WIN_C_STANDARD_LIBRARIES PREPEND "'")
+        list(FILTER WIN_C_STANDARD_LIBRARIES EXCLUDE REGEX "^''$")
```
Comment (strega-nil-ms) [inline scripts/cmake/vcpkg_configure_meson.cmake:93]:
```suggestion
        list(REMOVE_ITEM WIN_CXX_STANDARD_LIBRARIES "''")
```

File: scripts/cmake/vcpkg_configure_meson.cmake [hunk 11]
```diff
@@ -274,13 +287,24 @@ function(vcpkg_internal_meson_generate_cross_file _additional_binaries) #https:/
...
             string(APPEND CROSS "${proglower} = '${VCPKG_DETECTED_CMAKE_${prog}_COMPILER}'\n")
         endif()
     endforeach()
     if(VCPKG_DETECTED_CMAKE_LINKER AND VCPKG_TARGET_IS_WINDOWS)
-        string(APPEND CROSS "c_ld = '${VCPKG_DETECTED_CMAKE_LINKER}'\n")
-        string(APPEND CROSS "cpp_ld = '${VCPKG_DETECTED_CMAKE_LINKER}'\n")
+        if (NOT VCPKG_DETECTED_CMAKE_C_COMPILER MATCHES "gcc|icc")
+            string(APPEND CROSS "c_ld = '${VCPKG_DETECTED_CMAKE_LINKER}'\n")
+        elseif (NOT VCPKG_DETECTED_CMAKE_C_COMPILER STREQUAL VCPKG_DETECTED_CMAKE_LINKER)
+            message(FATAL_ERROR "The linker and compiler executeable must be the same for gcc and icc. Use -fuse-ld=... to change the linker backend.")
```
Comment (strega-nil-ms) [inline scripts/cmake/vcpkg_configure_meson.cmake:299]: same as above; s/executeable/executable and change VCPKG_DETECTED_CMAKE_C_COMPILER for VCPKG_DETECTED_CMAKE_C_COMPILER_ID.

File: scripts/cmake/vcpkg_configure_meson.cmake [hunk 12]
```diff
@@ -54,13 +54,24 @@ function(vcpkg_internal_meson_generate_native_file _additional_binaries) #https:
...
             string(REPLACE "CXX" "CPP" mesonprog "${prog}")
+            string(REPLACE "RC" "windres" mesonprog "${mesonprog}") # https://mesonbuild.com/Windows-module.html
             string(TOLOWER "${mesonprog}" proglower)
             string(APPEND NATIVE "${proglower} = '${VCPKG_DETECTED_CMAKE_${prog}_COMPILER}'\n")
         endif()
     endforeach()
     if(VCPKG_DETECTED_CMAKE_LINKER AND VCPKG_TARGET_IS_WINDOWS)
-        string(APPEND NATIVE "c_ld = '${VCPKG_DETECTED_CMAKE_LINKER}'\n")
-        string(APPEND NATIVE "cpp_ld = '${VCPKG_DETECTED_CMAKE_LINKER}'\n")
+        if (NOT VCPKG_DETECTED_CMAKE_C_COMPILER MATCHES "gcc|icc")
```
Comment (strega-nil-ms) [inline scripts/cmake/vcpkg_configure_meson.cmake:63]: I don't love this way of doing this; using the path of the compiler as a check for which compiler is being used is... shakey at best. Especially when you have things like cc as your VCPKG_DETECTED_CMAKE_C_COMPILER. Instead, `VCPKG_DETECTED_CMAKE_C_COMPILER_ID` should be used (and `CMAKE_C_COMPILER_ID` should be exported from `z_vcpkg_get_cmake_vars`) (and do the same for `VCPKG_DETECTED_CXX_COMPILER`)

```suggestion
        if (NOT VCPKG_DETECTED_CMAKE_C_COMPILER_ID MATCHES "^(GNU|Intel)$")
```

**Review by strega-nil-ms — CHANGES_REQUESTED** [review]
This is almost there; just needs slight changes.

The change in `get_cmake_vars` should be:

```diff
diff --git a/scripts/get_cmake_vars/CMakeLists.txt b/scripts/get_cmake_vars/CMakeLists.txt
index 1b2e1e561..b6d5f0e41 100644
--- a/scripts/get_cmake_vars/CMakeLists.txt
+++ b/scripts/get_cmake_vars/CMakeLists.txt
@@ -20,7 +20,7 @@ foreach(prog IN LISTS PROGLIST)
 endforeach()
 set(COMPILERS ${VCPKG_LANGUAGES} RC)
 foreach(prog IN LISTS COMPILERS)
-    list(APPEND VCPKG_DEFAULT_VARS_TO_CHECK CMAKE_${prog}_COMPILER)
+    list(APPEND VCPKG_DEFAULT_VARS_TO_CHECK CMAKE_${prog}_COMPILER CMAKE_${prog}_COMPILER_ID)^M
 endforeach()
 # Variables to check
 foreach(_lang IN LISTS VCPKG_LANGUAGES)
```

File: scripts/cmake/vcpkg_configure_meson.cmake [hunk 13]
```diff
@@ -121,11 +134,11 @@ function(vcpkg_internal_meson_generate_flags_properties_string _out_var _config)
     set(PATH_SUFFIX_DEBUG /debug)
     set(LIBPATH_${_config} "${L_FLAG}${CURRENT_INSTALLED_DIR}${PATH_SUFFIX_${_config}}/lib")
     vcpkg_internal_meson_convert_compiler_flags_to_list(MESON_CFLAGS_${_config} "${VCPKG_DETECTED_CMAKE_C_FLAGS_${_config}}")
-    list(APPEND MESON_CFLAGS_${_config} "-I\"${CURRENT_INSTALLED_DIR}/include\"")
+    list(APPEND MESON_CFLAGS_${_config} "-I${CURRENT_INSTALLED_DIR}/include")
```
Comment (Neumann-A) [inline scripts/cmake/vcpkg_configure_meson.cmake:137]: will meson pass this correctly if `CURRENT_INSTALLED_DIR` has a space ?

File: scripts/cmake/vcpkg_configure_meson.cmake (hunk 6 above)
Comment (autoantwort) [reply scripts/cmake/vcpkg_configure_meson.cmake]: Hm, the current solution does not work. The linker is kind of baked in the compiler, but it is still a different executable. So the current check fails because of the following paths are different. 
```
  VCPKG_DETECTED_CMAKE_C_COMPILER=/usr/bin/x86_64-w64-mingw32-gcc
  VCPKG_DETECTED_CMAKE_LINKER=/usr/bin/x86_64-w64-mingw32-ld
```

Comment (autoantwort) [reply scripts/cmake/vcpkg_configure_meson.cmake]: Hm it seems that `-fuse-ld` is a linker flag

File: scripts/cmake/vcpkg_configure_meson.cmake (hunk 13 above)
Comment (autoantwort) [reply scripts/cmake/vcpkg_configure_meson.cmake:137]: i wanted to test it on mac. But `bootstrap-vcpkg.sh` already fails with spaces in the cwd

Comment (autoantwort) [reply scripts/cmake/vcpkg_configure_meson.cmake:137]: It seems to work. So the argument 
```
cpp_args = ['-fPIC', '-g', '-I/Users/leanderSchulten/git_projekte/vcpkg test/vcpkg/installed/x64-osx/include']
```
 gets for example translated to
```
[3/6] /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/cc -Isrc/libepoxy.a.p -Isrc -I../src/1.5.5-5a2e9ec1d7.clean/src -Iinclude -I../src/1.5.5-5a2e9ec1d7.clean/include -Iinclude/epoxy '-I/Users/leanderSchulten/git_projekte/vcpkg test/vcpkg/installed/x64-osx/include' -fcolor-diagnostics -Wall -Winvalid-pch -std=gnu99 -g -fPIC -g -Wpointer-arith -Wmissing-declarations -Wformat=2 -Wstrict-prototypes -Wmissing-prototypes -Wnested-externs -Wbad-function-cast -Wold-style-definition -Wdeclaration-after-statement -Wunused -Wuninitialized -Wshadow -Wmissing-noreturn -Wmissing-format-attribute -Wredundant-decls -Werror=implicit -Werror=nonnull -Werror=init-self -Werror=main -Werror=missing-braces -Werror=sequence-point -Werror=return-type -Werror=trigraphs -Werror=array-bounds -Werror=write-strings -Werror=address -Werror=int-to-pointer-cast -Werror=pointer-to-int-cast -fno-strict-aliasing -Wno-int-conversion -MD -MQ src/libepoxy.a.p/dispatch_common.c.o -MF src/libepoxy.a.p/dispatch_common.c.o.d -o src/libepoxy.a.p/dispatch_common.c.o -c ../src/1.5.5-5a2e9ec1d7.clean/src/dispatch_common.c
```

File: scripts/cmake/vcpkg_configure_meson.cmake (hunk 6 above)
Comment (autoantwort) [reply scripts/cmake/vcpkg_configure_meson.cmake]: If I understand correctly you can not set the linker executable to something different for gcc. You have to use `-fuseld`. So we don't have to check if the linker executable was set to something different

Comment (BillyONeal) [reply scripts/cmake/vcpkg_configure_meson.cmake]: The point is that if the user has attempted to change the linker in the toolchain or triplet and we hit a place that can't be applied, we should fail with a message saying that the customization was not applied rather than silently succeeding with something the customer did not expect.

Comment (autoantwort) [reply scripts/cmake/vcpkg_configure_meson.cmake]: But how you would set a different linker in the toolchain/triplet? Afaik you have to set the `-fuseld` flag and flags should then be passed to cmake.

Comment (BillyONeal) [reply scripts/cmake/vcpkg_configure_meson.cmake]: > But how you would set a different linker in the toolchain/triplet? Afaik you have to set the `-fuseld` flag and flags should then be passed to meson.

To my understanding, they can change the linker by using `VCPKG_CHAINLOAD_TOOLCHAIN_FILE` ( https://github.com/microsoft/vcpkg/blob/master/docs/users/triplets.md#vcpkg_chainload_toolchain_file ) and doing anything that would influence cmake to choose a different linker.

Comment (autoantwort) [reply scripts/cmake/vcpkg_configure_meson.cmake]: In the toolchain you would then do something like 
```cmake
set(CMAKE_SHARED_LINKER_FLAGS_INIT "-fuse-ld=lld")
set(CMAKE_STATIC_LINKER_FLAGS_INIT "-fuse-ld=lld")
```
to change the linker for gcc. The flags gets detected and written to `VCPKG_DETECTED_CMAKE_XXX_LINKER_FLAGS` and then passed to meson.

File: scripts/cmake/vcpkg_configure_meson.cmake (hunk 2 above)
Comment (ras0219-msft) [reply scripts/cmake/vcpkg_configure_meson.cmake]: I think this really should be `VCPKG_CROSSCOMPILING` and the fact that always enabling this makes mingw work is actually papering over some other issues.

@dg0yt could you speak a bit more about what breaks with mingw meson? Even better would be a docker-based repro :)

**Review by ras0219-msft — COMMENTED** [review]
I'd really like to see `VCPKG_CROSSCOMPILING` used here -- let's make sure we've followed that down to root causes before merging.

**Comment by autoantwort** [comment]
> I'd really like to see `VCPKG_CROSSCOMPILING` used here -- let's make sure we've followed that down to root causes before merging.

Yeah [this](https://github.com/microsoft/vcpkg/pull/18927#issuecomment-881969583) breaks this currently, because we have to detect and pass the host c++ compiler for the non host triplet.

File: scripts/cmake/vcpkg_configure_meson.cmake [hunk 14]
```diff
@@ -54,13 +54,20 @@ function(vcpkg_internal_meson_generate_native_file _additional_binaries) #https:
...
-        string(APPEND NATIVE "c_ld = '${VCPKG_DETECTED_CMAKE_LINKER}'\n")
-        string(APPEND NATIVE "cpp_ld = '${VCPKG_DETECTED_CMAKE_LINKER}'\n")
+        if (NOT VCPKG_DETECTED_CMAKE_C_COMPILER_ID MATCHES "^(GNU|Intel)$") # for gcc and icc the linker flag -fuse-ld is used. See https://github.com/mesonbuild/meson/issues/8647#issuecomment-878673456
+            string(APPEND NATIVE "c_ld = '${VCPKG_DETECTED_CMAKE_LINKER}'\n")
+        endif()
+    endif()
+    if(VCPKG_DETECTED_CMAKE_LINKER AND VCPKG_TARGET_IS_WINDOWS)
+        if (NOT VCPKG_DETECTED_CMAKE_CXX_COMPILER_ID MATCHES "^(GNU|Intel)$") # for gcc and icc the linker flag -fuse-ld is used. See https://github.com/mesonbuild/meson/issues/8647#issuecomment-878673456
+            string(APPEND NATIVE "cpp_ld = '${VCPKG_DETECTED_CMAKE_LINKER}'\n")
+        endif()
```
Comment (Neumann-A) [inline scripts/cmake/vcpkg_configure_meson.cmake:70]: hmpf meson doing stupid stuff like:: 
```
defaults['static_linker'] = ['ar', 'gar']
defaults['strip'] = ['strip']
defaults['vs_static_linker'] = ['lib']
defaults['clang_cl_static_linker'] = ['llvm-lib']
defaults['cuda_static_linker'] = ['nvlink']
defaults['gcc_static_linker'] = ['gcc-ar']
defaults['clang_static_linker'] = ['llvm-ar']
```
and missing proper abstractions. 
see https://github.com/mesonbuild/meson/blob/master/mesonbuild/compilers/detect.py
I think we can completly remove the LINKER since there is no way to properly set it.

**Review by strega-nil-ms — APPROVED** [review]

**Comment by BillyONeal** [comment]
Thanks for your contribution!
